clang/test/CodeGenSYCL/no_opaque_virtual-types.cpp

```diff
@@ -20,9 +20,9 @@ int main() {
 // Struct layout big enough for vtable.
 // CHECK: %struct.Struct = type { i32 (...)** }
 // VTable:
-// CHECK: @_ZTV6Struct = linkonce_odr unnamed_addr constant { [3 x i8*] } { [3 x i8*] [i8* null, i8* bitcast ({ i8*, i8* }* @_ZTI6Struct to i8*), i8* bitcast (void (%struct.Struct addrspace(4)*)* @_ZN6Struct3fooEv to i8*)] }, comdat, align 8
-// CHECK: @[[TYPEINFO:.+]] = external global i8*
+// CHECK: @_ZTV6Struct = linkonce_odr unnamed_addr constant { [3 x i8*] } { [3 x i8*] [i8* null, i8* bitcast ({ i8 addrspace(4)*, i8* }* @_ZTI6Struct to i8*), i8* bitcast (void (%struct.Struct addrspace(4)*)* @_ZN6Struct3fooEv to i8*)] }, comdat, align 8
```

Comment:
Sorry, I didn't paid attention to it before. That's going to create a GV the private address space (under the SPIR convention). The SPIR-V translator will map that to `Function` storage class, which you can't use outside a function. So AFAIK, the generated SPIR-V won't be valid.

Out of curiosity, is there an extension you are working with here ? That would help to understand if the test is correct.

Reply:
> Out of curiosity, is there an extension you are working with here ? That would help to understand if the test is correct.

We don't have any SYCL-level extensions and for SPIR-V level we have [`SPV_INTEL_function_pointers`](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/spirv-extensions/SPV_INTEL_function_pointers.asciidoc)

Reply:
sorry, I meant a SPIR-V extension.

Looking at the function pointer ext, wouldn't this expression
`i8* bitcast (void (%struct.Struct addrspace(4)*)* @_ZN6Struct3fooEv to i8*)` be a problem ?  You can't cast to generic.

Reply:
Originally it was a problem described in your first comment - we were receiving `function` storage class RTTI global variables and hence llvm-translator crash. Now it fixed because we receiving `@_ZTVN10__cxxabiv117__class_type_infoE = external addrspace(1) global ptr addrspace(4)`, but not ` @_ZTVN10__cxxabiv117__class_type_infoE = external global ptr`.
I believe it's not a problem because applications run successfully.

Reply:
> I believe it's not a problem because applications run successfully.

well, it is not a problem for a specific stack, which is quite different.

I can see the `@[[TYPEINFO:.+]]` was indeed change the way you described it, and I think it make sense. But my comment for `_ZTV6Struct` (and `_ZTS6Struct` / `_ZTI6Struct`) still stand: this will generate invalid SPIR-V.

Reply:
> `i8* bitcast (void (%struct.Struct addrspace(4)*)* @_ZN6Struct3fooEv to i8*)`

@Naghasan, here we don't cast from generic to private, we cast a private pointer to one type to private pointer to another type. We have a function pointer `void(__generic Struct *)` which we cast to `i8*` without changing the address space of that function pointer

Reply:
You are right, I mixed up a few things in my brain by cross checking the test and the func ptr extension.

So in order to generate `void (%struct.Struct addrspace(4)*)* @_ZN6Struct3fooEv`  you need to use `OpConstantFunctionPointerINTEL` which yield a pointer to storage class `CodeSectionINTEL`  which you bitcast into a pointer to storage class `Function`. This is more what bugged me (I saw this by thinking of the implication of banning cast from generic storage class hence the confusion).

Strictly speaking, SPIR-V doesn't ban casting between storage class (at least I couldn't find wording banning this), which I found very strange (so does others, I'll seek clarification on that point).

Reply:
> So in order to generate void (%struct.Struct addrspace(4)*)* @_ZN6Struct3fooEv you need to use OpConstantFunctionPointerINTEL which yield a pointer to storage class CodeSectionINTEL which you bitcast into a pointer to storage class Function. This is more what bugged me (I saw this by thinking of the implication of banning cast from generic storage class hence the confusion).

In reality, we haven't implemented `CodeSectionINTEL` storage class or at least we haven't allocated a dedicated LLVM IR address space for it. For now it all just works with address spaces which are assigned by default to function pointers

Reply:
> we haven't implemented CodeSectionINTEL storage class

easy way to not hit the issue :)